### PR TITLE
Move some sync logic from rest to handlers pacakege

### PR DESCRIPTION
**Pull request by erikjohnston**
This includes PR https://github.com/matrix-org/synapse/pull/518

I've also changed it so that if a user gives an unrecognised filter id the server will return an error rather than using a default.

### CODE REVIEW

File: synapse/handlers/sync.py [hunk 1]
```diff
@@ -130,6 +131,11 @@ def __init__(self, hs):
         self.clock = hs.get_clock()
 
     @defer.inlineCallbacks
+    def get_sync_for_user(self, sync_config, since_token=None, timeout=0,
+                          filter_collection=DEFAULT_FILTER_COLLECTION):
```
Comment (NegativeMjark) [inline synapse/handlers/sync.py]: What is this?

File: synapse/handlers/sync.py [hunk 2]
```diff
@@ -684,21 +756,28 @@ def incremental_sync_with_gap_for_room(self, room_id, sync_config,
...
+        else:
+            state = {
+                (event.type, event.state_key): event
+                for event in batch.events if event.is_state()
+            }
 
         just_joined = yield self.check_joined_room(sync_config, state)
         if just_joined:
             state = yield self.get_state_at(room_id, now_token)
+            # batch.limited = True
```
Comment (NegativeMjark) [inline synapse/handlers/sync.py]: Should this be commented?

**Comment by erikjohnston** [comment]
PTAL

**Comment by erikjohnston** [comment]
matrixbot: test this please

**Comment by NegativeMjark** [comment]
LGTM

**Pull request merged**